python/cudf/cudf/core/column/column.py

```diff
@@ -1161,55 +1091,71 @@ def as_column(arbitrary, nan_as_null=True, dtype=None, name=None):
         data = arbitrary._values
         if dtype is not None:
             data = data.astype(dtype)
-    elif isinstance(arbitrary, Buffer):
-        data = numerical.NumericalColumn(data=arbitrary, dtype=arbitrary.dtype)
-
     elif isinstance(arbitrary, nvstrings.nvstrings):
-        data = string.StringColumn(data=arbitrary)
+        sbuf = Buffer.empty(arbitrary.byte_count())
+        obuf = Buffer.empty(
+            (arbitrary.size() + 1) * np.dtype("int32").itemsize
+        )
+
+        nbuf = None
+        if arbitrary.null_count() > 0:
+            mask_size = calc_chunk_size(arbitrary.size(), mask_bitsize)
+            nbuf = Buffer.empty(mask_size)
+            arbitrary.set_null_bitmask(nbuf.ptr, bdevmem=True)
+        arbitrary.to_offsets(sbuf.ptr, obuf.ptr, None, bdevmem=True)
+        children = (
+            build_column(obuf, dtype="int32"),
+            build_column(sbuf, dtype="int8"),
+        )
+        data = build_column(
+            data=None, dtype="object", mask=nbuf, children=children
+        )
+        data._nvstrings = arbitrary
+
+    elif isinstance(arbitrary, Buffer):
+        if dtype is None:
+            raise TypeError(f"dtype cannot be None if 'arbitrary' is a Buffer")
+        data = build_column(arbitrary, dtype=dtype)
 
     elif cuda.devicearray.is_cuda_ndarray(arbitrary):
```

Comment:
do we even need this separately from handling `__cuda_array_interface__` now?

Reply:
I think it's useful to keep the two cases distinct. Buffer does have a `__cuda_array_interface__`, but the `type` metadata there is useless. To construct a Column from a Buffer, we *must* provide a `dtype` whereas the same is not true for `__cuda_array_interface__`.

Reply:
I was referring to handling the Numba DeviceArray object that we check for here with `cuda.devicearray.is_cuda_ndarray(...)`

Reply:
Sorry - the UI doesn't make it super clear which LOC(s) a review comment refers to.

It looks like we have some special handling of DeviceArrays where we convert NaNs to nulls, whereas in the `__cuda_array_interface__` case we just use the value of the `"mask"`

Reply:
Should we convert NaNs to nulls from `__cuda_array_interface__` where the mask doesn't exist? That would cover inputs like CuPy, Numba, Deep Learning Frameworks, etc.

Reply:
Is there still something that needs to be done here?

Reply:
Decided to do this in a followup PR, will raise an issue after this is merged.